crates/kumo-dmarc/src/lib.rs

```diff
@@ -0,0 +1,498 @@
+use chrono::{DateTime, Utc};
+use instant_xml::{FromXml, ToXml};
+use kumo_spf::SpfDisposition;
+use std::fmt::{self, Write};
+use std::net::IpAddr;
+use std::str::FromStr;
+
+enum Format {
+    Afrf,
+}
+
+impl FromStr for Format {
+    type Err = String;
+
+    fn from_str(value: &str) -> Result<Self, Self::Err> {
+        Ok(match value {
+            "afrf" => Self::Afrf,
+            _ => return Err(format!("invalid format {value:?}")),
+        })
+    }
+}
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq, ToXml)]
+#[xml(scalar)]
+enum Policy {
+    None,
+    Quarantine,
+    Reject,
+}
+
+impl FromStr for Policy {
+    type Err = String;
+
+    fn from_str(value: &str) -> Result<Self, Self::Err> {
+        Ok(match value {
+            "none" => Self::None,
+            "quarantine" => Self::Quarantine,
+            "reject" => Self::Reject,
+            _ => return Err(format!("invalid policy {value:?}")),
+        })
+    }
+}
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq, ToXml)]
+#[xml(scalar)]
+enum Mode {
+    Relaxed,
+    Strict,
+}
+
+impl From<Mode> for char {
+    fn from(value: Mode) -> Self {
+        match value {
+            Mode::Relaxed => 'r',
+            Mode::Strict => 's',
+        }
+    }
+}
+
+impl FromStr for Mode {
+    type Err = String;
+
+    fn from_str(value: &str) -> Result<Self, Self::Err> {
+        Ok(match value {
+            "r" => Self::Relaxed,
+            "s" => Self::Strict,
+            _ => return Err(format!("invalid mode {value:?}")),
+        })
+    }
+}
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+struct ReportFailure {
+    all_pass: bool,
+    any_pass: bool,
+    dkim: bool,
+    spf: bool,
+}
+
+impl ToXml for ReportFailure {
+    fn serialize<W: std::fmt::Write + ?Sized>(
+        &self,
+        _: Option<instant_xml::Id<'_>>,
+        serializer: &mut instant_xml::Serializer<W>,
+    ) -> Result<(), instant_xml::Error> {
+        serializer.write_str(self)
+    }
+}
+
+impl fmt::Display for ReportFailure {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let values = [
+            (self.all_pass, '0'),
+            (self.any_pass, '1'),
+            (self.dkim, 'd'),
+            (self.spf, 's'),
+        ];
+
+        let mut first = true;
+        for &(_, ch) in values.iter().filter(|(value, _)| *value) {
```

Comment:
I feel like this is a bit easier to grok at first sight:

```suggestion
        for (value, ch) in values {
            if !value {
                continue;
             }
```

Reply:
Fair enough, if I was reviewing this code I might've said the same thing! Will fix it.